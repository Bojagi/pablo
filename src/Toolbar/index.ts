export * from './Toolbar';
export * from './ToolbarItem';
export * from './ToolbarDivider';
