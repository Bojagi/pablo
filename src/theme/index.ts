export * from './defaultTheme';
export * from './context';
export * from './PabloThemeProvider';
export * from './useComponentStyle';
