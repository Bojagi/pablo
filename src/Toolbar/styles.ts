import { getColor } from '../utils/styleHelpers/getColor';
import { getSpacing } from '../utils/styleHelpers/getSpacing';
import { Style } from '../theme/types';

export interface ToolbarStyles {
  gap: Style;
  divider: {
    width: number;
    color: Style;
  };
  item: {
    size: Style;
    borderRadius: number;
    backgroundColor: Style;
    color: Style;
    buttonTransition: string[][];
    iconTransition: string[][];
    iconScale: number;
    hover: {
      backgroundColor: Style;
      color: Style;
    };
    focus: {
      backgroundColor: Style;
      color: Style;
    };
    active: {
      iconScale: number;
      backgroundColor: Style;
      color: Style;
    };
  };
}

export const toolbarStyles: ToolbarStyles = {
  gap: getSpacing(2),
  divider: {
    width: 1,
    color: getColor('borders', 'lightest'),
  },
  item: {
    buttonTransition: [
      ['background-color', '0.2s'],
      ['color', '0.2s'],
    ],
    iconTransition: [['transform', '0.2s']],
    iconScale: 1,
    size: '24px',
    borderRadius: 6,
    backgroundColor: 'transparent',
    color: getColor('common', 'black'),
    hover: {
      backgroundColor: getColor('brand', 'lightest'),
      color: getColor('common', 'black'),
    },
    focus: {
      backgroundColor: getColor('brand', 'lightest'),
      color: getColor('common', 'black'),
    },
    active: {
      iconScale: 0.833333333,
      backgroundColor: getColor('brand'),
      color: getColor('common', 'white'),
    },
  },
};
