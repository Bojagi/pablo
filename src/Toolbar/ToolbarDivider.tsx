import styled from 'styled-components';
import { baseStyle } from '../shared/baseStyle';
import { getComponentStyle } from '../styleHelpers/getComponentStyle';
import { BaseProps } from '../types';
import { getCustomStyles } from '../utils/useCustomStyles';
import { ToolbarDividerStyleProperties } from './styles';

export const ToolbarDivider = styled('div').attrs({
  role: 'separator',
  'aria-orientation': 'vertical',
})<BaseProps<ToolbarDividerStyleProperties>>`
  ${baseStyle}
  margin: 0 ${getComponentStyle('toolbar.gap')};
  width: ${getComponentStyle('toolbar.divider.width')}px;
  align-self: stretch;
  background-color: ${getComponentStyle('toolbar.divider.color')};
  ${getCustomStyles('toolbar.divider.styles', 'root')}
`;
