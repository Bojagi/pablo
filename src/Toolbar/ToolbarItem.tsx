import React, { forwardRef } from 'react';
import { LayoutBoxProps } from '../Box';
import { IconButton } from '../IconButton';
import { useComponentStyle } from '../theme/useComponentStyle';
import { Tooltip, TooltipSide } from '../Tooltip/Tooltip';
import { BaseProps } from '../types';
import { ToolbarItemStyleProperties } from './styles';

export interface ToolbarItemProps
  extends Omit<LayoutBoxProps, 'size'>,
    BaseProps<ToolbarItemStyleProperties> {
  active?: boolean;
  icon: React.ReactNode;
  disabled?: boolean;
  tooltip?: React.ReactNode;
  tooltipSide?: TooltipSide;
  onClick?: (name: string) => void;
  name: string;
}

export const ToolbarItem = forwardRef<HTMLButtonElement, ToolbarItemProps>(
  ({ active, tooltip, tooltipSide, icon, name, onClick, customStyles, ...props }, ref) => {
    const gap = useComponentStyle('toolbar.gap');
    const themeCustomStyles: any = useComponentStyle('toolbar.item.styles');
    const combinedCustomStyles: any = {
      ...themeCustomStyles,
      ...customStyles,
    };

    return (
      <Tooltip delay={500} side={tooltipSide} content={tooltip} disabled={props.disabled}>
        <IconButton
          mx={gap}
          ref={ref}
          data-testid="pbl-toolbar-item-button"
          active={active}
          onClick={() => onClick && onClick(name)}
          customStyles={combinedCustomStyles}
          {...props}
          size="small"
        >
          {icon}
        </IconButton>
      </Tooltip>
    );
  }
);
