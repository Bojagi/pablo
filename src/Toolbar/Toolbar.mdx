---
name: Toolbar
route: /toolbar
menu: Components
---

import { Props, Playground } from 'docz'
import { Box } from '../Box/Box.tsx'
import { Toolbar } from './Toolbar.tsx'
import { ToolbarItem } from './ToolbarItem.tsx'
import { ToolbarDivider } from './ToolbarDivider.tsx'
import { Bold, Underline, Italic, Edit, Crop, CornerUpLeft } from 'react-feather';

# Toolbar

A `Toolbar` is a set of `IconButtons` that can be used for actions or states of a specific function or section 
(e.g. rich text editors or a drawing canvas).

## Examples

### Simple Toolbar

Below you can see a simple toolbar. The `selected` state for the `ToolbarItem`s is specified as a `Toolbar` prop.
The `ToolbarItem`s have a name, that is passed as first argument in the `onClick` function, so you can just pass a
hook setter function. 

<Playground>
  {() => {
    const [selectedTool, setSelectedTool] = React.useState('bold');
    return (
      <Toolbar selected={selectedTool}>
        <ToolbarItem
          name="bold"
          onClick={setSelectedTool}
          icon={<Bold size={20} />}
        />
        <ToolbarItem
          name="underline"
          onClick={setSelectedTool}
          icon={<Underline size={20} />}
        />
        <ToolbarItem
          name="italic"
          onClick={setSelectedTool}
          icon={<Italic size={20} />}
        />
      </Toolbar>
    );
  }}
</Playground>

### Tooltips

It's advices do use a `tooltip`, so the toolbar is more accessable and easier to understand. 
In case the normal tooltip position covers or overlaps with other content you can specify a side where the tooltip is show.
The side value can be any valid `Tooltip` `side` prop value.

<Playground>
  {() => {
    const [selectedTool, setSelectedTool] = React.useState('bold');
    return (
      <Toolbar selected={selectedTool}>
        <ToolbarItem
          name="bold"
          tooltip="Make text bold"
          tooltipSide="left"
          onClick={setSelectedTool}
          icon={<Bold size={20} />}
        />
        <ToolbarItem
          name="underline"
          tooltip="Underline text"
          onClick={setSelectedTool}
          icon={<Underline size={20} />}
        />
        <ToolbarItem
          name="italic"
          tooltip="Make text italic"
          tooltipSide="right"
          onClick={setSelectedTool}
          icon={<Italic size={20} />}
        />
      </Toolbar>
    );
  }}
</Playground>

### Dividing a Toolbar

In case you have subsections for your toolbar, you can divide them with a bar.

<Playground>
  {() => {
    const [selectedTool, setSelectedTool] = React.useState('bold');
    return (
      <Toolbar selected={selectedTool}>
        <ToolbarItem
          name="bold"
          onClick={setSelectedTool}
          icon={<Bold size={20} />}
        />
        <ToolbarItem
          name="underline"
          onClick={setSelectedTool}
          icon={<Underline size={20} />}
        />
        <ToolbarItem
          name="italic"
          onClick={setSelectedTool}
          icon={<Italic size={20} />}
        />
        <ToolbarDivider />
        <ToolbarItem
          name="edit"
          onClick={setSelectedTool}
          icon={<Edit size={20} />}
        />
        <ToolbarItem
          name="crop"
          onClick={setSelectedTool}
          icon={<Crop size={20} />}
        />
      </Toolbar>
    );
  }}
</Playground>

### Mixing tool selection and one time actions

Toolbars can mix single tool selection where only one item can be selected at a time and one time action buttons (like undo).
This can be done by setting the `onClick` prop with a different setter or action function. In the example below you can
also see `ToolbarItem`s being disabled (by clicking on the undo button).
<Playground>
  {() => {
    const [selectedTool, setSelectedTool] = React.useState('bold');
    const [undoDisabled, setUndoDisabled] = React.useState(false);
    return (
      <Toolbar selected={selectedTool}>
        <ToolbarItem
          name="bold"
          onClick={setSelectedTool}
          icon={<Bold size={20} />}
        />
        <ToolbarItem
          name="underline"
          onClick={setSelectedTool}
          icon={<Underline size={20} />}
        />
        <ToolbarItem
          name="italic"
          onClick={setSelectedTool}
          icon={<Italic size={20} />}
        />
        <ToolbarDivider />
        <ToolbarItem
          name="undo"
          disabled={undoDisabled}
          onClick={() => setUndoDisabled(true)}
          icon={<CornerUpLeft size={20} />}
        />
      </Toolbar>
    );
  }}
</Playground>

## Import

```js
import { Toolbar, ToolbarItem, ToolbarDivider } from '@bojagi/pablo/Toolbar';
```

## Props

<Props of={Toolbar} />
