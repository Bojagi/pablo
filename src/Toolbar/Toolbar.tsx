import React, { forwardRef, cloneElement } from 'react';
import styled from '@emotion/styled';
import { layoutInterpolationFn, LayoutBoxProps } from '../Box';
import { guaranteeArray } from '../utils/guaranteeArray';
import { getComponentStyle } from '../styleHelpers';
import { ToolbarItem, ToolbarItemProps } from './ToolbarItem';
import { getCustomStyles } from '../utils/useCustomStyles';
import { BaseProps } from '../types';
import { ToolbarStyleProperties } from './styles';
import { baseStyle } from '../shared/baseStyle';

const ToolbarBox = styled.div<LayoutBoxProps>`
  ${baseStyle}
  display: flex;
  margin: 0 -${getComponentStyle(['toolbar', 'gap'])};
  height: calc(${getComponentStyle(['toolbar', 'item', 'size'])} + 2 * ${getComponentStyle(['toolbar', 'gap'])});
  align-items: center;
  ${layoutInterpolationFn}
  ${getCustomStyles('toolbar.styles', 'root')}
`;

export interface ToolbarProps extends LayoutBoxProps, BaseProps<ToolbarStyleProperties> {
  children:
    | React.ReactElement<ToolbarItemProps | object>
    | React.ReactElement<ToolbarItemProps | object>[];
  selected?: string;
}

export const Toolbar = forwardRef<HTMLDivElement, ToolbarProps>(
  ({ children, selected, ...props }: ToolbarProps, ref) => {
    const interpolatedChildren = guaranteeArray(children).map((child, index) => {
      if (child.type === ToolbarItem) {
        const childProps = child.props as ToolbarItemProps;

        return cloneElement(child, {
          key: childProps.name,
          active: selected ? childProps.name === selected : childProps.active,
        });
      }

      return cloneElement(child, { key: index });
    });
    return (
      <ToolbarBox role="toolbar" aria-orientation="horizontal" ref={ref} {...props}>
        {interpolatedChildren}
      </ToolbarBox>
    );
  }
);
