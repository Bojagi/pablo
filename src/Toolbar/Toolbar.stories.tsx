import React, { useState } from 'react';

import { Bold, Italic, Crop, Edit, Underline, ArrowLeft } from 'react-feather';
import { css } from '@emotion/react';
import { Toolbar } from './Toolbar';
import { ToolbarDivider } from './ToolbarDivider';
import { ToolbarItem } from './ToolbarItem';

export default {
  title: 'Toolbar',
};

export const SimpleToolbar = () => {
  const [selectedTool, setSelectedTool] = useState('bold');
  const [undoDisabled, setUndoDisabled] = useState(false);
  return (
    <Toolbar my={9} mx={10} selected={selectedTool}>
      <ToolbarItem
        name="bold"
        tooltip="Make text bold"
        tooltipSide="left"
        onClick={setSelectedTool}
        icon={<Bold size={20} />}
      />
      <ToolbarItem
        name="underline"
        tooltip="Underline text"
        onClick={setSelectedTool}
        icon={<Underline size={20} />}
      />
      <ToolbarItem
        name="italic"
        tooltip="Make text italic"
        tooltipSide="bottom"
        onClick={setSelectedTool}
        icon={<Italic size={20} />}
      />
      <ToolbarDivider />
      <ToolbarItem
        name="edit"
        tooltip="Edit the item"
        onClick={setSelectedTool}
        icon={<Edit size={20} />}
      />
      <ToolbarItem
        name="crop"
        tooltip="Crop image"
        tooltipSide="right"
        onClick={setSelectedTool}
        icon={<Crop size={20} />}
      />
      <ToolbarItem
        name="undo"
        tooltip="Undo"
        tooltipSide="right"
        disabled={undoDisabled}
        onClick={() => setUndoDisabled(true)}
        icon={<ArrowLeft size={20} />}
      />
    </Toolbar>
  );
};

const customToolbarItemStyles = {
  root: css`
    background-color: red;
  `,
  hover: css`
    background-color: orange;
  `,
  active: css`
    background-color: purple;
  `,
};

export const CustomStyledToolbar = () => {
  const [selectedTool, setSelectedTool] = useState('bold');
  const [undoDisabled, setUndoDisabled] = useState(false);

  return (
    <Toolbar
      my={9}
      mx={10}
      selected={selectedTool}
      customStyles={{
        root: css`
          border: 1px solid blue;
        `,
      }}
    >
      <ToolbarItem
        name="bold"
        tooltip="Make text bold"
        tooltipSide="left"
        onClick={setSelectedTool}
        icon={<Bold size={20} />}
        customStyles={customToolbarItemStyles}
      />
      <ToolbarItem
        name="underline"
        tooltip="Underline text"
        onClick={setSelectedTool}
        icon={<Underline size={20} />}
        customStyles={customToolbarItemStyles}
      />
      <ToolbarItem
        name="italic"
        tooltip="Make text italic"
        tooltipSide="bottom"
        onClick={setSelectedTool}
        icon={<Italic size={20} />}
        customStyles={customToolbarItemStyles}
      />
      <ToolbarDivider
        customStyles={{
          root: css`
            transform: rotate(45deg);
            background-color: yellow;
          `,
        }}
      />
      <ToolbarItem
        name="edit"
        tooltip="Edit the item"
        onClick={setSelectedTool}
        icon={<Edit size={20} />}
        customStyles={customToolbarItemStyles}
      />
      <ToolbarItem
        name="crop"
        tooltip="Crop image"
        tooltipSide="right"
        onClick={setSelectedTool}
        icon={<Crop size={20} />}
        customStyles={customToolbarItemStyles}
      />
      <ToolbarItem
        name="undo"
        tooltip="Undo"
        tooltipSide="right"
        disabled={undoDisabled}
        onClick={() => setUndoDisabled(true)}
        icon={<ArrowLeft size={20} />}
        customStyles={customToolbarItemStyles}
      />
    </Toolbar>
  );
};
