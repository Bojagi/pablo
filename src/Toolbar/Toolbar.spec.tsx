import { act, fireEvent, render } from '@testing-library/react';
import React from 'react';
import { waitOneTick } from '../../testUtils/waitOneTick';
import { PabloThemeProvider } from '../theme';
import { Toolbar } from './Toolbar';
import { ToolbarDivider } from './ToolbarDivider';
import { ToolbarItem } from './ToolbarItem';
import '../../testUtils/mockResizeObserver';
import { themeVars } from '../theme/themeVars';

beforeEach(() => {
  jest.useFakeTimers();
});

test('Render toolbar', () => {
  const { container } = renderComponent({}, {});
  expect(container).toMatchSnapshot();
});

test('Render toolbar with one item active', async () => {
  const { getAllByTestId } = render(
    <PabloThemeProvider>
      <Toolbar>
        <ToolbarItem name="bold" icon={<div>bold-icon</div>} />
        <ToolbarItem name="underline" active icon={<div>underline-icon</div>} />
        <ToolbarItem name="italic" icon={<div>italic-icon</div>} />
        <ToolbarDivider />
        <ToolbarItem name="edit" icon={<div>edit-icon</div>} />
        <ToolbarItem name="crop" icon={<div>crop-icon</div>} />
      </Toolbar>
    </PabloThemeProvider>
  );
  // This is because of passing ref on effect, which happens on next tick
  await act(() => Promise.resolve());

  const allButtons = getAllByTestId('pbl-toolbar-item-button');
  expect(allButtons).toBeArrayOfSize(5);
  expect(allButtons[0]).toHaveStyleRule('background-color', 'transparent');
  expect(allButtons[1]).toHaveStyleRule('background-color', themeVars.colors.brand.main);
  expect(allButtons[2]).toHaveStyleRule('background-color', 'transparent');
  expect(allButtons[3]).toHaveStyleRule('background-color', 'transparent');
  expect(allButtons[4]).toHaveStyleRule('background-color', 'transparent');
});

test('Render toolbar with all items active', async () => {
  const { getAllByTestId } = renderComponent({}, { active: true });
  // This is because of passing ref on effect, which happens on next tick
  await act(() => Promise.resolve());

  const allButtons = getAllByTestId('pbl-toolbar-item-button');
  expect(allButtons).toBeArrayOfSize(6);
  expect(allButtons[0]).toHaveStyleRule('background-color', themeVars.colors.brand.main);
  expect(allButtons[1]).toHaveStyleRule('background-color', themeVars.colors.brand.main);
  expect(allButtons[2]).toHaveStyleRule('background-color', themeVars.colors.brand.main);
  expect(allButtons[3]).toHaveStyleRule('background-color', themeVars.colors.brand.main);
  expect(allButtons[4]).toHaveStyleRule('background-color', themeVars.colors.brand.main);
  expect(allButtons[5]).toHaveStyleRule('background-color', themeVars.colors.brand.main);
});

test('Render toolbar with disabled button disabled', async () => {
  const { getAllByTestId } = renderComponent({}, {});
  // This is because of passing ref on effect, which happens on next tick
  await act(() => Promise.resolve());

  const allButtons = getAllByTestId('pbl-toolbar-item-button');
  expect(allButtons[0]).not.toHaveAttribute('disabled');
  expect(allButtons[1]).not.toHaveAttribute('disabled');
  expect(allButtons[2]).not.toHaveAttribute('disabled');
  expect(allButtons[3]).not.toHaveAttribute('disabled');
  expect(allButtons[4]).not.toHaveAttribute('disabled');
  expect(allButtons[5]).toHaveAttribute('disabled');
});

test('Render toolbar items selected by "selected" Toolbar prop', async () => {
  const { getAllByTestId } = renderComponent({ selected: 'italic' }, { active: true });
  // This is because of passing ref on effect, which happens on next tick
  await act(() => Promise.resolve());

  const allButtons = getAllByTestId('pbl-toolbar-item-button');
  expect(allButtons).toBeArrayOfSize(6);
  expect(allButtons[0]).toHaveStyleRule('background-color', 'transparent');
  expect(allButtons[1]).toHaveStyleRule('background-color', 'transparent');
  expect(allButtons[2]).toHaveStyleRule('background-color', themeVars.colors.brand.main);
  expect(allButtons[3]).toHaveStyleRule('background-color', 'transparent');
  expect(allButtons[4]).toHaveStyleRule('background-color', 'transparent');
});

test('Trigger onClick on all items and call function with item name as argument', async () => {
  const onClickMock = jest.fn();
  const { getAllByTestId } = renderComponent({}, { onClick: onClickMock });
  // This is because of passing ref on effect, which happens on next tick
  await act(() => Promise.resolve());

  const allButtons = getAllByTestId('pbl-toolbar-item-button');
  expect(onClickMock).toHaveBeenCalledTimes(0);

  // Click "bold" item
  fireEvent.click(allButtons[0]);
  expect(onClickMock).toHaveBeenCalledTimes(1);
  expect(onClickMock).toHaveBeenCalledWith('bold');

  // Click "underline" item
  fireEvent.click(allButtons[1]);
  expect(onClickMock).toHaveBeenCalledTimes(2);
  expect(onClickMock).toHaveBeenCalledWith('underline');

  // Click "italic" item
  fireEvent.click(allButtons[2]);
  expect(onClickMock).toHaveBeenCalledTimes(3);
  expect(onClickMock).toHaveBeenCalledWith('italic');

  // Click "edit" item
  fireEvent.click(allButtons[3]);
  expect(onClickMock).toHaveBeenCalledTimes(4);
  expect(onClickMock).toHaveBeenCalledWith('edit');

  // Click "crop" item
  fireEvent.click(allButtons[4]);
  expect(onClickMock).toHaveBeenCalledTimes(5);
  expect(onClickMock).toHaveBeenCalledWith('crop');
});

test('Render toolbar items with tooltip', async () => {
  const { getAllByTestId } = renderComponent({}, { tooltip: 'i am a tooltip' });
  // This is because of passing ref on effect, which happens on next tick
  await act(() => Promise.resolve());

  const allButtons = getAllByTestId('pbl-toolbar-item-button');

  act(() => {
    allButtons.forEach((btn) => {
      fireEvent.mouseEnter(btn);
    });
  });

  act(() => {
    jest.advanceTimersByTime(500);
  });

  await waitOneTick(0, true);

  const allTooltips = getAllByTestId('pbl-tooltip-popover');
  expect(allTooltips).toBeArrayOfSize(5);
  expect(allTooltips[0]).toHaveTextContent('i am a tooltip');
  expect(allTooltips[1]).toHaveTextContent('i am a tooltip');
  expect(allTooltips[2]).toHaveTextContent('i am a tooltip');
  expect(allTooltips[3]).toHaveTextContent('i am a tooltip');
  expect(allTooltips[4]).toHaveTextContent('i am a tooltip');
});

function renderComponent(toolbarProps, toolbarItemProps) {
  return render(
    <PabloThemeProvider>
      <Toolbar {...toolbarProps}>
        <ToolbarItem name="bold" {...toolbarItemProps} icon={<div>bold-icon</div>} />
        <ToolbarItem name="underline" {...toolbarItemProps} icon={<div>underline-icon</div>} />
        <ToolbarItem name="italic" {...toolbarItemProps} icon={<div>italic-icon</div>} />
        <ToolbarDivider />
        <ToolbarItem name="edit" {...toolbarItemProps} icon={<div>edit-icon</div>} />
        <ToolbarItem name="crop" {...toolbarItemProps} icon={<div>crop-icon</div>} />
        <ToolbarItem
          name="disabled"
          disabled
          {...toolbarItemProps}
          icon={<div>disabled-icon</div>}
        />
      </Toolbar>
    </PabloThemeProvider>
  );
}
