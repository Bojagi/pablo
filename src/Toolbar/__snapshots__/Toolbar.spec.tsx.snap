// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Render toolbar 1`] = `
.emotion-0 {
  box-sizing: border-box;
  display: -webkit-box;
  display: -webkit-flex;
  display: -ms-flexbox;
  display: flex;
  margin: 0 -clamp(0.25rem, 0.2142857142857143rem + 0.17857142857142858vw, 0.375rem);
  height: calc(24px + 2 * clamp(0.25rem, 0.2142857142857143rem + 0.17857142857142858vw, 0.375rem));
  -webkit-align-items: center;
  -webkit-box-align: center;
  -ms-flex-align: center;
  align-items: center;
}

.emotion-1 {
  box-sizing: border-box;
  width: 24px;
  height: 24px;
  background-color: transparent;
  color: var(--pbl-theme-colors-common-black);
  border: 0;
  padding: 0;
  outline: 0;
  box-sizing: border-box;
  -webkit-transition: background-color 0.2s,color 0.2s;
  transition: background-color 0.2s,color 0.2s;
  border-radius: 6px;
  display: -webkit-box;
  display: -webkit-flex;
  display: -ms-flexbox;
  display: flex;
  -webkit-box-pack: center;
  -ms-flex-pack: center;
  -webkit-justify-content: center;
  justify-content: center;
  -webkit-align-items: center;
  -webkit-box-align: center;
  -ms-flex-align: center;
  align-items: center;
  -webkit-flex-shrink: 0;
  -ms-flex-negative: 0;
  flex-shrink: 0;
  margin-left: clamp(0.25rem, 0.2142857142857143rem + 0.17857142857142858vw, 0.375rem);
  margin-right: clamp(0.25rem, 0.2142857142857143rem + 0.17857142857142858vw, 0.375rem);
}

.emotion-1>* {
  width: 20px;
  height: 20px;
  -webkit-transition: -webkit-transform 0.2s;
  transition: transform 0.2s;
  -webkit-transform: scale(
        1
      );
  -moz-transform: scale(
        1
      );
  -ms-transform: scale(
        1
      );
  transform: scale(
        1
      );
}

.emotion-1:enabled {
  cursor: pointer;
}

.emotion-1:disabled {
  opacity: 0.3;
  cursor: normal;
}

.emotion-1:hover:enabled {
  background-color: var(--pbl-theme-colors-blackOpacity-50);
  color: var(--pbl-theme-colors-common-black);
}

.emotion-1:focus:enabled {
  background-color: var(--pbl-theme-colors-blackOpacity-100);
  color: var(--pbl-theme-colors-common-black);
}

.emotion-4 {
  box-sizing: border-box;
  margin: 0 clamp(0.25rem, 0.2142857142857143rem + 0.17857142857142858vw, 0.375rem);
  width: 1px;
  -webkit-align-self: stretch;
  -ms-flex-item-align: stretch;
  align-self: stretch;
  background-color: var(--pbl-theme-colors-borders-light);
}

<div>
  <div
    aria-orientation="horizontal"
    class="emotion-0"
    role="toolbar"
  >
    <button
      aria-pressed="false"
      class="emotion-1"
      color="plain"
      data-testid="pbl-toolbar-item-button"
      role="toolbar"
    >
      <div>
        bold-icon
      </div>
    </button>
    <button
      aria-pressed="false"
      class="emotion-1"
      color="plain"
      data-testid="pbl-toolbar-item-button"
      role="toolbar"
    >
      <div>
        underline-icon
      </div>
    </button>
    <button
      aria-pressed="false"
      class="emotion-1"
      color="plain"
      data-testid="pbl-toolbar-item-button"
      role="toolbar"
    >
      <div>
        italic-icon
      </div>
    </button>
    <div
      aria-orientation="vertical"
      class="emotion-4"
      role="separator"
    />
    <button
      aria-pressed="false"
      class="emotion-1"
      color="plain"
      data-testid="pbl-toolbar-item-button"
      role="toolbar"
    >
      <div>
        edit-icon
      </div>
    </button>
    <button
      aria-pressed="false"
      class="emotion-1"
      color="plain"
      data-testid="pbl-toolbar-item-button"
      role="toolbar"
    >
      <div>
        crop-icon
      </div>
    </button>
    <button
      aria-pressed="false"
      class="emotion-1"
      color="plain"
      data-testid="pbl-toolbar-item-button"
      disabled=""
      role="toolbar"
    >
      <div>
        disabled-icon
      </div>
    </button>
  </div>
</div>
`;
